### ISSUE TITLE
Steam controller bug

### ISSUE THREAD

**Issue by ashuntu** [canonical/steam-snap#280]
### Discussed in https://github.com/canonical/steam-snap/discussions/268

<div type='discussions-op-text'>

<sup>Originally posted by **UnrealisticDino** June 25, 2023</sup>
Steam does not recognize that the controller is connected which means I can't remap inputs. With the adapter the controller does work mostly as expected. In Mighty No. 9 and 10 Second Ninja touching up or right on the left touch pad acts like holding right on the left analog stick until the touch pad is touched again on either the left or down areas. In A Story About My Uncle the right trigger doesn't work. If the controller is in Bluetooth mode the controller does not change inputs when in game and instead keeps behaving like a keyboard and mouse. </div>

**Comment by UnrealisticDino** [comment]
This bug is about the steam controller.

**Comment by ashuntu** [comment]
@UnrealisticDino is this still an issue with the current version of the Snap? We should have fixed a huge portion of controller problems. I happen to have a Steam controller too (with the included USB dongle) and have been able to use it just fine in Apex Legends, including changing the controller layout.

If it's no longer an issue, please close this issue. If it still is, make sure `joystick` and `uinput` plugs are connected, failing that, try installing `steam-devices`, e.g. `sudo apt install --no-install-recommends steam-devices`.

**Comment by UnrealisticDino** [comment]
I'll try to report back by tomorrow, but if I don't I will by the end of a week.

**Comment by UnrealisticDino** [comment]
> @UnrealisticDino is this still an issue with the current version of the Snap? We should have fixed a huge portion of controller problems. I happen to have a Steam controller too (with the included USB dongle) and have been able to use it just fine in Apex Legends, including changing the controller layout.
> 
> If it's no longer an issue, please close this issue. If it still is, make sure `joystick` and `uinput` plugs are connected, failing that, try installing `steam-devices`, e.g. `sudo apt install --no-install-recommends steam-devices`.

I'm having the same issue on my pc and just to double check did a clean install of ubuntu 22.04 on my laptop. I did run sudo apt install --no-install-recommends steam-devices on both even though I already had steam devices installed on my pc. Use any connected joystick and uinput: allows access to the uinput device are both enabled and are enabled by default. Installing steam devices used to fix this issue which is why I said when only the snap version of steam is installed. I really should have been more clear about that looking back.

**Comment by ashuntu** [comment]
Does having the Steam deb installed still "fix" the issue for you?

**Comment by UnrealisticDino** [comment]
No. I remember that having the deb version installed made it so that the snap version could recognize that the steam controller was plugged in but now it shows that no controller is detected. I also remember that just having the deb installed fixed it but the deb did not have to be running. There are bigger issues to be fixed so don't mind if this is not a priority.

**Comment by blain1972** [comment]
I was unable to get Steam to recognise my steam controller at all - although I was able to use to navigate around steam itself. I've been trying to get this working on and off for a few days. I had the relevant plugs connected to the snap and steam-devices installed.

After reading this I installed the .deb and I was able to connect it straight away. 

I've uninstalled the snap for now.

**Comment by vincenttschanz** [comment]
I just installed Ubuntu 24.10 on a new PC and tried the Snap version of Steam. The Steam controller is still not detected on Steam (USB or wireless dongle) . Installing steam-devices didn't help. 

I installed the Deb version and it detected the controller directly.

**Comment by ashuntu** [comment]
This issue is a bit tough to debug since our team does not own any Steam controllers. However, here are the general things you should try and see if they fix any issues. If not, then it's useful to note what you've tried and what works/doesn't. Note: these aren't always individual fixes, sometimes it may be useful to try combinations of them.

- Try a snapd version *greater* than 2.63 (at the moment, this is `candidate` or newer).
    ```
    snap refresh snapd --edge
    ```
    *Make sure to revert after testing*
    ```
    snap refresh snapd --stable
    ```
- Make sure `uinput` and `joystick` are connected (if you really want to test these, it may be best to connect *all* interfaces and see if anything changes first, and progressively remove them to see which affect behavior)
    ```
    snap connect steam:uninput
    snap connect steam:joystick
    ```
- Try the `edge` channel of the Steam snap (not relevant at the time of writing, but sometimes changes are made that may affect things)
    ```
    snap refresh steam --edge
    ```
    *Make sure to revert after testing*
    ```
    snap refresh steam --stable
    ```
- Try installing `steam-devices`
    ```
    apt install --no-install-recommends steam-devices
    ```
    - Alternatively, install the Steam deb alongside the snap (it provides some external tools in addition to Steam, like `steam-devices`)
- Try the controller in the Steam deb to see if it works (based on feedback, it seems like it almost always does)

And as always, [logs](https://github.com/canonical/steam-snap/wiki/Troubleshooting#view-logs) can sometimes be helpful.
